go/workflow.go

```diff
@@ -0,0 +1,333 @@
+	if customer.StatusLevel == nil {
+		customer.StatusLevel = StatusLevelForPoints(customer.LoyaltyPoints)
+	}
+	if len(customer.Guests) == 0 {
+		customer.Guests = make([]string, customer.StatusLevel.GuestsAllowed)
+	}
+}
+
+func addGuestToCustomer(customer *CustomerInfo, guestID string) error {
+	guestExists := false
+	for _, g := range customer.Guests {
+		guestExists = g == guestID
+	}
+	// replace first empty slot, or append if none
+	if !guestExists {
+		added := false
+		for i, g := range customer.Guests {
+			if len(g) == 0 {
+				customer.Guests[i] = guestID
+				added = true
+				break
+			}
+		}
+		if !added {
+			customer.Guests = append(customer.Guests, guestID)
+		}
+	}
+
+	return nil
+}
+
+func signalAddPoints(ctx workflow.Context, c workflow.ReceiveChannel, customer *CustomerInfo) error {
+	logger := workflow.GetLogger(ctx)
+	activities := Activities{}
+
+	var pointsToAdd int
+	c.Receive(ctx, &pointsToAdd)
+
+	logger.Info("Adding points to customer account.", "PointsAdded", pointsToAdd)
+	customer.LoyaltyPoints += pointsToAdd
+
+	currentStatusOrd := customer.StatusLevel.Ordinal
+	customer.StatusLevel = StatusLevelForPoints(customer.LoyaltyPoints)
+	newStatusOrd := customer.StatusLevel.Ordinal
+
+	statusChange := newStatusOrd - currentStatusOrd
+
+	if statusChange > 0 {
+		err := workflow.ExecuteActivity(ctx, activities.SendEmail,
+			fmt.Sprintf(EmailPromoted, customer.StatusLevel.Name)).
+			Get(ctx, nil)
+		if err != nil {
+			return fmt.Errorf("error running SendEmail activity for status promotion: %w", err)
+		}
+	} else if statusChange < 0 {
+		err := workflow.ExecuteActivity(ctx, activities.SendEmail,
+			fmt.Sprintf(EmailDemoted, customer.StatusLevel.Name)).
+			Get(ctx, nil)
+		if err != nil {
+			return fmt.Errorf("error running SendEmail activity for status demotion: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func signalInviteGuest(ctx workflow.Context, c workflow.ReceiveChannel, customer *CustomerInfo) error {
+	logger := workflow.GetLogger(ctx)
+	activities := Activities{}
+
+	var emailToSend string
+	var guestID string
+	c.Receive(ctx, &guestID)
+
+	logger.Info("Checking to see if customer has enough status to allow for a guest invite.", "Customer", customer)
+	nonEmptyGuests := 0
+	for _, v := range customer.Guests {
+		if len(v) > 0 {
+			nonEmptyGuests++
+		}
+	}
+	if nonEmptyGuests < customer.StatusLevel.GuestsAllowed {
+		logger.Info("Customer is allowed to invite guests. Attempting to invite.",
+			"GuestID", guestID)
+
+		guest := CustomerInfo{
+			CustomerID:    guestID,
+			StatusLevel:   customer.StatusLevel.Previous(),
+			AccountActive: true,
+		}
+
+		err := addGuestToCustomer(customer, guestID)
+		if err != nil {
+			return fmt.Errorf("unable to add guest '%v' to customer's guest list: %w", guestID, err)
+		}
+
+		err = workflow.ExecuteLocalActivity(ctx, activities.StartGuestWorkflow, guest).Get(ctx, nil)
+
+		t := &GuestAlreadyCanceledError{}
+		if errors.As(err, &t) {
```

Comment:
Hrmm, I think this only works because it's a local activity. This wouldn't work in a normal activity (errors are not serialized back to their original type). I fear users will see this and think it's an ok thing to do w/ normal activity errors. I think we should either use application error and check the type, or not use error and use some return value instead.